extension-compose/src/main/java/com/mapbox/maps/extension/compose/internal/MapApplier.kt

```diff
@@ -0,0 +1,83 @@
+package com.mapbox.maps.extension.compose.internal
+
+import androidx.compose.runtime.AbstractApplier
+import com.mapbox.maps.MapView
+
+/**
+ * Defines the contract of a MapNode, the MapNode will be notified when it's added/moved/removed from
+ * the node tree.
+ */
+internal interface MapNode {
+  /**
+   * Invoked when the [MapNode] is attached to the MapboxMap node tree.
+   */
+  fun onAttached() {}
+
+  /**
+   * Invoked when the [MapNode] is moved inside the node tree.
+   *
+   * @param from original position
+   * @param to the target position
+   */
+  fun onMoved(from: Int, to: Int) {}
+
+  /**
+   * Invoked when the [MapNode] is removed from the MapboxMap node tree.
+   */
+  fun onRemoved() {}
+
+  /**
+   * Invoked when the MapboxMap node tree is cleared.
+   */
+  fun onClear() {}
+}
+
+/**
+ * Root level [MapNode] for MapboxMap composable function.
+ */
+private object RootMapNode : MapNode
+
+/**
+ * MapApplier is responsible for applying the tree-based operations on MapboxMap, that get emitted
+ * during a composition.
+ */
+internal class MapApplier(
+  /**
+   * The raw map controller to be used within current Compose.
+   */
+  val mapView: MapView
+) : AbstractApplier<MapNode>(RootMapNode) {
+  private val decorations = mutableListOf<MapNode>()
+
+  override fun onClear() {
+    decorations.forEach { it.onClear() }
+    decorations.clear()
+  }
+
+  override fun insertBottomUp(index: Int, instance: MapNode) {
+    decorations.add(index, instance)
+    instance.onAttached()
```

Comment:
should we notify another mapNodes about such kind of adding? maybe another mapNode will be fully hidden?

Reply:
other mapNode are independent on the newly added node
Newly added node will be placed on the top of existing nodes.